source/IpcModule.ts

```diff
@@ -0,0 +1,57 @@
+import { ipcRenderer, ipcMain } from 'electron'
+
+type Promisified<T extends Function> = 
+    T extends (...args: infer P) => infer R
+        ? ( R extends Promise<any> 
+            ? (...args: P) => R
+            : (...args: P) => Promise<R>
+          )
+        : never
+    ;
+
+type Ipcfied<D> = {
+        [K in keyof D]: D[K] extends Function ? Promisified<D[K]> : never;
+    }
+
+export abstract class IpcModule {
+    public static forRenderer<T>(): Ipcfied<T> {
+        return new Proxy({}, {
+            get (target: any, prop: PropertyKey, receiver: any): any {
+                const methodName = prop.toString();
+                return async function(...args: any[]) {
+                    return ipcRenderer.invoke(methodName, args);
+                }
+            }
+        });
+    }
+
+    public static registerMain<T>(service: T) {
+        // Get all methods of the service
+        const userFunctions = IpcModule.getAllUserFunctions(service)
```

Comment:
Do I get this right that you simply extract EVERYTHING that can serve as a getter from the class? Isn't that leaking unintentional functionality as well? o_O

Reply:
Yes, it exposes all methods of the service. If this is too much, then the caller can restrict it by using
```
registerMain(
 new class implements T {
    //...only implement the required methods by forwarding them to wherever you want
})
```